src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -461,27 +428,21 @@ void BaseCouplingScheme::checkCompletenessRequiredActions()
   }
 }
 
-void BaseCouplingScheme::setupDataMatrices(DataMap &data)
+void BaseCouplingScheme::setupDataMatrices()
 {
   PRECICE_TRACE();
-  PRECICE_DEBUG("Data size: {}", data.size());
+  PRECICE_DEBUG("Data size: {}", getAccelerationData().size());
   // Reserve storage for convergence measurement of send and receive data values
   for (ConvergenceMeasureContext &convMeasure : _convergenceMeasures) {
     PRECICE_ASSERT(convMeasure.couplingData != nullptr);
-    if (convMeasure.couplingData->oldValues.cols() < 1) {
-      utils::append(convMeasure.couplingData->oldValues,
-                    (Eigen::MatrixXd) Eigen::MatrixXd::Zero(convMeasure.couplingData->values().size(), 1));
-    }
+    convMeasure.couplingData->storeIteration();
   }
-  // Reserve storage for extrapolation of data values
-  if (_extrapolationOrder > 0) {
-    for (DataMap::value_type &pair : data) {
-      int cols = pair.second->oldValues.cols();
-      PRECICE_DEBUG("Add cols: {}, cols: {}", pair.first, cols);
-      PRECICE_ASSERT(cols <= 1, cols);
-      utils::append(pair.second->oldValues,
-                    (Eigen::MatrixXd) Eigen::MatrixXd::Zero(pair.second->values().size(), _extrapolationOrder + 1 - cols));
-    }
+  // Reserve storage for data values
+  for (DataMap::value_type &pair : getAccelerationData()) {
+    time::PtrWaveform       ptrWaveform(new time::Waveform(pair.second->values().size(), _extrapolationOrder + 2));
```

Comment:
Why "extrapolationOrder + 2"? Why not just "extrapolationOrder"? Otherwise, this becomes hard to understand.

Reply:
1. It should be a `extrapolationOrder + 1`, since we do the extrapolation in-place.
2. For zeroth order extrapolation you need one column (to read from and write to), for first order two, for second order three... Currently, we can also handle this inside of the `Waveform`. The parameter currently defines how much "memory" the waveform needs. 

Btw: zeroth order extrapolation is a special case. If we want to do linear interpolation inside the window in the future, we will also need two columns (beginning and end of window). But for the actual extrapolation one one column at the end of the window is needed.

Reply:
Fixed this in 24e7d63e9aebdb19484b11f95e2ab7373a1cff84